src/Compiler/Microsoft.CodeAnalysis.Razor.Compiler/src/Language/DefaultRazorProjectItem.cs

```diff
@@ -9,36 +9,58 @@ namespace Microsoft.AspNetCore.Razor.Language;
 
 internal class DefaultRazorProjectItem : RazorProjectItem
 {
+    private FileInfo _fileInfo;
+    private readonly string _physicalFilePath;
     private readonly string _fileKind;
 
     /// <summary>
     /// Initializes a new instance of <see cref="DefaultRazorProjectItem"/>.
     /// </summary>
     /// <param name="basePath">The base path.</param>
-    /// <param name="relativePhysicalPath">The physical path of the base path.</param>
     /// <param name="filePath">The path.</param>
+    /// <param name="relativePhysicalPath">The physical path of the base path.</param>
     /// <param name="fileKind">The file kind. If null, the document kind will be inferred from the file extension.</param>
     /// <param name="file">The <see cref="FileInfo"/>.</param>
     /// <param name="cssScope">A scope identifier that will be used on elements in the generated class, or null.</param>
     public DefaultRazorProjectItem(string basePath, string filePath, string relativePhysicalPath, string fileKind, FileInfo file, string cssScope)
     {
         BasePath = basePath;
         FilePath = filePath;
+        _physicalFilePath = null;
+        RelativePhysicalPath = relativePhysicalPath;
+        _fileKind = fileKind;
+        _fileInfo = file;
+        CssScope = cssScope;
+    }
+
+    /// <summary>
+    /// Initializes a new instance of <see cref="DefaultRazorProjectItem"/>.
+    /// </summary>
+    /// <param name="basePath">The base path.</param>
+    /// <param name="filePath">The path.</param>
+    /// <param name="physicalPath">The physical path of the file path.</param>
+    /// <param name="relativePhysicalPath">The physical path of the base path.</param>
+    /// <param name="fileKind">The file kind. If null, the document kind will be inferred from the file extension.</param>
+    /// <param name="cssScope">A scope identifier that will be used on elements in the generated class, or null.</param>
+    public DefaultRazorProjectItem(string basePath, string filePath, string physicalPath, string relativePhysicalPath, string fileKind, string cssScope)
+    {
+        BasePath = basePath;
+        FilePath = filePath;
+        _physicalFilePath = physicalPath;
         RelativePhysicalPath = relativePhysicalPath;
         _fileKind = fileKind;
-        File = file;
         CssScope = cssScope;
     }
 
-    public FileInfo File { get; }
+    public FileInfo File => _fileInfo ??= new(FilePath);
```

Comment:
Should the FileInfo be created from `_physicalFilePath`?

Reply:
Previously, the `FileInfo` object created by `DefaultRazorProjectFileSystem.GetItem(...)` was created with the string that gets stored in `FilePath`. The `FileInfo` objects created by `DefaultRazorProjectFileSystem.EnumerateItems(...)` are likely created with `FullPath`. When making `FileInfo` lazily created on `DefaultRazorProjectItem`, I found that no Razor product code actually calls `EnumerateItems(...)`, but there are calls to `GetItem(...)`. So, I chose to preserve the behavior of `GetItem(...)`.

FWIW, nothing in Razor accesses `DefaultRazorProjectItem.File` either, so it's possible that we could just remove that property and do away with the `FileInfo` altogether.

Reply:
Okay, it's fine that `File` is created from `FilePath`.

What's more concerning is that `Exists` and `PhysicalPath` properties below might change depending on whether someone has accessed `File` or not:

```cs
var x = new DefaultRazorProjectItem(...);
x.PhysicalPath // returns _physicialPath
x.File // _fileInfo is created from FilePath
x.PhysicialPath // returns _fileInfo.FullName, i.e., FilePath now
```

In other words, perhaps they should use `FilePath` instead of `_physicalFilePath`?

Reply:
Yeah, I agree that it's all too subtle and poorly defined. `RazorProjectItem.PhysicalFilePath` is supposed to point to the real file system path. It's created by accessing `FileInfo.FullPath` or calling `Path.GetFullPath(...)`. `RazorProjectItem.FilePath` is constructed by taking `RelativePhysicalPath` (the path from the root to the file), swapping all slashes to `/` and adding a `/` at the start.

I'm strongly considering just removing the `DefaultRazorProjectItem.File` property. It doesn't look like anything actually accesses this property in this repo or dotnet/sdk. Would that help?

Note that nothing significant actually accesses `Exists`. That could be removed too. :smile: